# Convolutional Neural Net
- **CNN** is basically two dimensional configuration of neural networks
- The input of CNN are image (three N by N if it color image and N by N if its black and white image)
- The weights are also two dimensional array

**CNN Weights** are known as Kernels or Filter matrix <br>
**Stride** is the name for its' filter-matrix/kernals horizontal and vertical movement <br>
    (Stride visualization: http://deeplearning.stanford.edu/wiki/index.php/Feature_extraction_using_convolution) <br>
**Output Size** = (input_size - filter_size)/stride + 1

### CNN Weights 
Are known as Kernals or Filter Matricies <br>
![CNN Weights](../Images/kernel_image.png)

### Stride
The name for the CNNs' filter-matricies/kernals horizontal and vertical movements <br>

Stride Equal To One
![Stride Example One](../Images/stride_1.png) <br>
Stride Equalt To Two
![Stride Example Two](../Images/stride_2.png) <br>


- Stride visualization: http://deeplearning.stanford.edu/wiki/index.php/Feature_extraction_using_convolution <br>
- **Output Size** = (input_size - filter_size)/stride + 1

### Max Pooling
**Max pooling:** take the maximum element from each window of a certain size <br>
![Max Pooling Image](../Images/maxpooling.png)

### Faltten Layer
- After feature extraction that is done by multiple Convolutional layers, **we use flatten layer to add MLP after convolutional layers in order to do classification task**
- This one is simple--it's just Keras's version of numpy.reshape. This reshapes n-dimensional arrays to a vector. This is necessary when moving from Conv2D layers, which expect 2-dimensional arrays as inputs, to Dense layers, which expect 1-dimension vectors as inputs. As a concrete example, a Flatten layer given a 28 x 28 array as input would output a vector of the shape (784, 1)
<br>
<br>

## Visualizing a Neural Net (CNN + MLP
![Neural Net Process Image](../Images/CNN.png)

## Activity: Obtain the number of parameters for the following CNN
The default strides = (1, 1)

In [4]:
from __future__ import print_function
import keras

from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D


model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(28, 28, 1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
# model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
# model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

print(model.summary())

Using TensorFlow backend.


Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               1179776   
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 1,199,882
Trainable params: 1,199,882
Non-trainable params: 0
____________________________________________

**output_size** = (28 - 3)/1 + 1 = 26 <br>
**output_size** = (26 - 3)/1 + 1 = 24 <br>
**First Conv2D Parameters:** = 32 x 9 + 32 = 320 <br>
**Second Conv2D Parameters:** = 64 x 32 x 9 + 64 = 18496 <br>
**Flatten Shape:** 12 x 12 x 64 = 9216 <br>
**Dense One Parameters** = 9216 x 128 + 128 = 1179776 <br>
**Dense Two Parameters** = 128 x 10 + 10 = 1290 <br>

## Activity: Explore CNN visualizations and explain what's happening
Find Visualization Here: https://www.cs.ryerson.ca/~aharley/vis/conv/flat.html

## Data Preparation for CNN
Suppose we want to feed a 4 by 4 image to a CNN network, how we should reshape the data?

In [5]:
from keras.layers import Input, Dense, Conv2D, MaxPooling2D, UpSampling2D
from keras.models import Model
from keras import backend as K
import numpy as np

input_img = Input(shape=(4, 4, 1))  # adapt this if using `channels_first` image data format

x = Conv2D(2, (2, 2), activation='relu')(input_img)
y = Conv2D(3, (2, 2), activation='relu')(x)
model = Model(input_img, y)
# cnv_ml_1 = Model(input_img, x)

data = np.array([[5, 12, 1, 8], [2, 10, 3, 6], [4, 7, 9, 1], [5, 7, 5, 6]])
data = data.reshape(1, 4, 4, 1)
print(model.predict(data))
print('M :')
print(model.predict(data).reshape(3, 2, 2))
print(model.summary())

[[[[0.        0.        8.790258 ]
   [3.603326  6.9580956 2.6464162]]

  [[0.        0.        7.820859 ]
   [2.005819  3.0757465 6.913532 ]]]]
M :
[[[0.        0.       ]
  [8.790258  3.603326 ]]

 [[6.9580956 2.6464162]
  [0.        0.       ]]

 [[7.820859  2.005819 ]
  [3.0757465 6.913532 ]]]
Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 4, 4, 1)           0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 2)           10        
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 2, 2, 3)           27        
Total params: 37
Trainable params: 37
Non-trainable params: 0
_________________________________________________________________
None


In [6]:
from keras.datasets import mnist

img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

print(x_train[0].shape)
print(x_train[1].shape)

11493376/11490434 [==============================] - 7s 1us/step
(28, 28, 1)
(28, 28, 1)
